src/lib.rs

```diff
@@ -73,21 +75,22 @@ mod platform;
 /// type into a buffer.
 pub trait CopyAddress {
     /// Copy an address into user-defined buffer.
-    fn copy_address(&self, addr: usize, buf: &mut [u8]) -> std::io::Result<()>;
+    fn copy_address(&self, addr: usize, buf: &mut Vec<u8>) -> std::io::Result<()>;
 
     /// Get the actual memory location from a set of offsets.
     ///
     /// If [`copy_address`] is already defined, then we can provide a standard implementation that
     /// will work across all operating systems.
-    fn get_offset(&self, offsets: &[usize]) -> std::io::Result<usize> {
+    fn get_offset(&self, offsets: &[usize], arch: &usize) -> std::io::Result<usize> {
         // Look ma! No unsafes!
         let mut offset: usize = 0;
         let noffsets: usize = offsets.len();
         for next_offset in offsets.iter().take(noffsets - 1) {
             offset += next_offset;
-            let mut copy: [u8; std::mem::size_of::<usize>()] = [0; std::mem::size_of::<usize>()];
+            let mut copy: Vec<u8> = vec![0; arch / 8];
             self.copy_address(offset, &mut copy)?;
-            offset = usize::from_ne_bytes(copy);
+            copy.resize(std::mem::size_of::<usize>(), 0);
+            offset = usize::from_ne_bytes(copy.as_slice().try_into().unwrap());
```

Comment:
I'm not a massive fan of this section of code.

Adding a `/` to here will slow things down, as will the allocation of a new vector every iteration. If we do want a vector to be used, then I'd suggest putting it outside loop and using `set_len()` or `resize()` prior to the call to `copy_address()`.

You're also assuming [endianness](https://en.wikipedia.org/wiki/Endianness) here. `from_ne_bytes()` converts from native-endian ordered bytes to a type, and by simply padding a vector with 0s, you can potentially end up with the wrong number if we have a big-endian system.

One potential fix for this would be to just read the full `usize`, and then mask. This would look something like the following:

```rust
let mut copy: [u8; std::mem::size_of::<usize>()] = [0; std::mem::size_of::<usize>()];
self.copy_address(offset, &mut copy)?;
offset = usize::from_ne_bytes(copy) & mask;
```
where mask is `0x0000ffff` for 32-bit and `0xffffffff` for 64-bit and is a usize.

Reply:
I agree with you on the vector point.

I've been looking at the masking and I realise that both my initial approach, and the masking approaches cease to work if the size of pointers in the target process is larger than the size of `usize` since the `offset` variable may not be able to store the pointers.  While that can be easily changed it would be nice to support arbitrary size input offset lists too, which (I think from my limited Rust knowledge) would mean making the `DataMember` type generic on the pointer size?

Do you have any other suggestions for how to solve this?

Reply:
I don't think there's an elegant solution to the case where we want to have target pointers larger than the host pointer size. I strongly suspect that the OS APIs that are in use wouldn't support such a thing anyway - most of them take a host pointer as the address to read from a remote address.

Reply:
Perhaps just implement for pointer sizes that are smaller than the host size? I can't think of any applications that would run a (for example) 64-bit app inside a 32-bit host without virtualization of a different CPU.